packages/ckeditor5-typing/src/utils/delete/injectbeforeinputdeletehandling.js

```diff
@@ -0,0 +1,104 @@
+ * on {@link module:engine/view/document~Document view document} and executes the `delete` or `forwardDelete` commands
+ * in web browsers that support Input Events (`beforeinput`).
+ *
+ * @param {module:core/editor/editor~Editor} editor The editor instance.
+ */
+export default function injectBeforeInputDeleteHandling( editor ) {
+	const editingView = editor.editing.view;
+	const viewDocument = editingView.document;
+
+	viewDocument.on( 'delete', ( evt, data ) => {
+		const { direction, sequence, selectionToRemove, inputType, unit } = data;
+
+		// Both "deleteContentBackward" and "deleteContentForward" must operate on the unit-level despite the
+		// editing view range available and, in case of other input types, used efficiently to delete content.
+		// This is related to the multi-byte characters decomposition (like complex emojis). Check out
+		// the comments in DeleteObserver's DELETE_EVENT_TYPES to learn more.
+		if ( inputType === 'deleteContentBackward' ) {
+			const deleteCommandData = {
+				sequence,
+				unit
+			};
+
+			// Android IMEs have a quirk which is addressed by passing a complete selection in case of
+			// the 'deleteContentBackward' event type. See DeleteObserver#_enableBeforeInputBasedObserver()
+			// to learn more.
+			if ( env.isAndroid ) {
+				deleteCommandData.selection = selectionToRemove;
+			}
+
+			editor.execute( 'delete', deleteCommandData );
+		} else if ( inputType === 'deleteContentForward' ) {
+			editor.execute( 'forwardDelete', {
+				unit,
+				sequence
+			} );
+		}
+		// In case of other delete (beforeinput) types, use the range provided by the beforeinput event.
+		else {
+			const modelRanges = [ ...selectionToRemove.getRanges() ].map( viewRange => {
+				return editor.editing.mapper.toModelRange( viewRange );
+			} );
+			const selection = editor.model.createSelection( modelRanges );
+			const isForwardDelete = direction === 'forward';
+
+			editor.execute( isForwardDelete ? 'forwardDelete' : 'delete', {
+				selection,
+				sequence
+			} );
+		}
+
+		editingView.scrollToTheSelection();
+	} );
+
+	// Android IMEs have a quirk - they change DOM selection after the input changes were performed by the browser.
+	// This happens on `keyup` event. Android doesn't know anything about our deletion and selection handling.
+	// Even if the selection was changed during input events, IME remembers the position where the selection "should"
+	// be placed and moves it there.
+	//
+	// To prevent incorrect selection, we save the selection after deleting here and then re-set it on `keyup`.
+	// This has to be done on DOM selection level, because on `keyup` the model selection is still the same as it was
+	// just after deletion, so it wouldn't be changed and the fix would do nothing.
+	//
+	// **Note**: See DeleteObserver#_enableBeforeInputBasedObserver() for the first part of this quirk.
+	if ( env.isAndroid ) {
```

Comment:
I don't like this code either but this would result in `DeleteObserver` listening to its own event (`#delete`), which is against the single-responsibility principle. And it would look weird 😛 TBH I'm not even 100% sure it works at this moment. I'm not saying no, but let's get back to this code in #8011 maybe?

Reply:
Ok.